Add additional labels/annotations, and add unit tests for constructJobsFromTemplate

danielvegamyhre opened the pull request:
Fixes #35 
Fixes #30

- On every job created from a replicated job, add a label storing the total replica count.
- Add job index as a label and annotation to the pod template spec.
- Refactor `constructJobsFromTemplate` so it is more easily testable and add several unit tests for it.

ahg-g commented on api/v1alpha1/jobset_types.go:
```diff
@@ -20,6 +20,7 @@ import (
 const (
 	JobIndexLabel string = "jobset.sigs.k8s.io/job-index"
+	ReplicasLabel string = "jobset.sigs.k8s.io/job-replicas"
 	RestartsLabel string = "jobset.sigs.k8s.io/restart-attempt"
```
lets call them keys, not labels

ahg-g commented on pkg/controllers/jobset_controller.go:
```diff
@@ -384,14 +384,23 @@ func (r *JobSetReconciler) constructJob(js *jobset.JobSet, rjob *jobset.Replicat
+	job.Spec.Template.Labels = cloneMap(job.Spec.Template.Labels)
+	job.Spec.Template.Annotations = cloneMap(job.Spec.Template.Annotations)
+	job.Spec.Template.Labels[jobset.JobIndexLabel] = strconv.Itoa(jobIdx)
+	job.Spec.Template.Annotations[jobset.JobIndexLabel] = strconv.Itoa(jobIdx)
```
add the replicas too; this can be used by initContainers for jobs like MPI

ahg-g commented on pkg/controllers/jobset_controller.go:
```diff
@@ -384,14 +384,23 @@ func (r *JobSetReconciler) constructJob(js *jobset.JobSet, rjob *jobset.Replicat
 
-	// Add restart-attempt count label, it should be equal to jobSet restarts
-	// to indicate is part of the current jobSet run.
+	// Add replica count as label to the job.
```
lets add unit tests for `constructJobsFromTemplate`; to make it easy to test, we can remove it from being a member function of `JobSetReconciler`

danielvegamyhre commented:
@ahg-g Thanks for the review, I addressed your comments and refactored `constructJobsFromTemplate` so it is more easily testable, and added several unit tests for it.

ahg-g commented on pkg/controllers/jobset_controller.go:
```diff
@@ -336,21 +341,52 @@ func (r *JobSetReconciler) deleteJobs(ctx context.Context, js *jobset.JobSet, jo
+	return nil
+}
+
+// TODO: update condition in place if it exists.
```
we should do this in a follow up PR, this is important

danielvegamyhre replied:
Created #39 and referenced it in this TODO.

ahg-g commented on pkg/controllers/jobset_controller_test.go:
```diff
@@ -263,3 +265,305 @@ func TestSetExclusiveAffinities(t *testing.T) {
+	exclusive *jobset.Exclusive
+}
+
+func makeJob(args *expectedJobArgs) *batchv1.Job {
```
I would make this return a job wrapper and remove the subdomain and exclusive args since each are used in a single test.

For example, the subdomain test will look like: `makeJob().Subdomain("test-jobset-replicated-job-0")`; same with the exclusive test.

danielvegamyhre replied:
Makes sense, done.

ahg-g commented:
/label tide/merge-method-squash

ahg-g commented:
/lgtm
